File: go/quota.go
package statapi

import (
	"net/http"
	"strconv"
)

// Quota is a snapshot of the caller's monthly record quota, parsed from the
// X-Quota-* headers stamped on every API response. Present is false when the
// response carried no quota headers; the counts are zero in that case.
type Quota struct {
	Limit     int64
	Used      int64
	Remaining int64
	Present   bool
}

func parseQuota(h http.Header) Quota {
	limit, okLimit := int64Header(h, "X-Quota-Limit")
	used, okUsed := int64Header(h, "X-Quota-Used")
	remaining, okRemaining := int64Header(h, "X-Quota-Remaining")
	if !okLimit && !okUsed && !okRemaining {
		return Quota{}
	}
	return Quota{Limit: limit, Used: used, Remaining: remaining, Present: true}
}

func int64Header(h http.Header, name string) (int64, bool) {
	raw := h.Get(name)
	if raw == "" {
		return 0, false
	}
	n, err := strconv.ParseInt(raw, 10, 64)
	if err != nil {
		return 0, false
	}
	return n, true
}

File: dotnet/src/Quota.cs
namespace StatApi;

/// <summary>
/// Snapshot of the caller's monthly record quota.
/// </summary>
/// <remarks>
/// Hand-written core (not generated). Parsed from the <c>X-Quota-*</c> response
/// headers stamped on every API response; any component the server omits is
/// <c>null</c>.
/// </remarks>
public sealed class Quota
{
    public long? Limit { get; }
    public long? Used { get; }
    public long? Remaining { get; }

    public Quota(long? limit, long? used, long? remaining)
    {
        Limit = limit;
        Used = used;
        Remaining = remaining;
    }

    internal static Quota? FromHeaders(HttpResponseMessage response)
    {
        var limit = Read(response, "x-quota-limit");
        var used = Read(response, "x-quota-used");
        var remaining = Read(response, "x-quota-remaining");
        if (limit is null && used is null && remaining is null)
        {
            return null;
        }
        return new Quota(limit, used, remaining);
    }

    private static long? Read(HttpResponseMessage response, string name)
    {
        if (response.Headers.TryGetValues(name, out var values))
        {
            foreach (var value in values)
            {
                if (long.TryParse(value, out var parsed))
                {
                    return parsed;
                }
            }
        }
        return null;
    }
}
